validator/client/src/main/java/tech/pegasys/teku/validator/client/StableSubnetSubscriber.java

```diff
@@ -0,0 +1,151 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.validator.client;
+
+import static java.lang.Integer.min;
+import static tech.pegasys.teku.util.config.Constants.ATTESTATION_SUBNET_COUNT;
+import static tech.pegasys.teku.util.config.Constants.EPOCHS_PER_RANDOM_SUBNET_SUBSCRIPTION;
+import static tech.pegasys.teku.util.config.Constants.RANDOM_SUBNETS_PER_VALIDATOR;
+import static tech.pegasys.teku.util.config.Constants.SLOTS_PER_EPOCH;
+
+import com.google.common.primitives.UnsignedLong;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.NavigableSet;
+import java.util.Optional;
+import java.util.Random;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.stream.IntStream;
+import tech.pegasys.teku.validator.api.SubnetSubscription;
+import tech.pegasys.teku.validator.api.ValidatorApiChannel;
+
+public class StableSubnetSubscriber {
+
+  private final ValidatorApiChannel validatorApiChannel;
+  private final Set<Integer> availableSubnetIndices = new HashSet<>();
+  private final NavigableSet<SubnetSubscription> subnetSubscriptions =
+      new TreeSet<>(
+          Comparator.comparing(SubnetSubscription::getUnsubscriptionSlot)
+              .thenComparing(SubnetSubscription::getSubnetId));
+  private final Random random;
+
+  private volatile int validatorCount;
+
+  public StableSubnetSubscriber(
+      ValidatorApiChannel validatorApiChannel, Random random, int validatorCount) {
+    this.validatorApiChannel = validatorApiChannel;
+    this.validatorCount = validatorCount;
+    this.random = random;
+    IntStream.range(0, ATTESTATION_SUBNET_COUNT).forEach(availableSubnetIndices::add);
+  }
+
+  public void onSlot(UnsignedLong slot) {
+    // Iterate through current subscriptions to remove the ones that have expired
+    final Iterator<SubnetSubscription> iterator = subnetSubscriptions.iterator();
+    while (iterator.hasNext()) {
+      final SubnetSubscription subnetSubscription = iterator.next();
+      if (subnetSubscription.getUnsubscriptionSlot().compareTo(slot) > 0) {
+        break;
+      }
+
+      iterator.remove();
+      int subnetId = subnetSubscription.getSubnetId();
+      availableSubnetIndices.add(subnetId);
+    }
+
+    // Adjust the number of subscriptions
+    // If any update was made to the subscriptions pass the new subscription set to BeaconNode
+    if (adjustNumberOfSubscriptionsToNumberOfValidators(slot, validatorCount)) {
+      validatorApiChannel.subscribeToPersistentSubnets(subnetSubscriptions);
+    }
+  }
+
+  public void updateValidatorCount(final int validatorCount) {
```

Comment:
Does this get called anywhere outside of tests?

Reply:
No, but the idea was that it might we might implement validator exits and such soon, so it would be good to have such a mechanism ready.